packages/vite/src/node/build.ts

```diff
@@ -1252,3 +1318,172 @@ function areSeparateFolders(a: string, b: string) {
+  run: () => Promise<void>
+  cancel: () => void
+}
+
+export interface BuilderOptions {
+  runBuildTasks?: (
+    builder: ViteBuilder,
+    buildTasks: BuildTask[],
+  ) => Promise<void>
+}
+
+async function defaultRunBuildTasks(
+  builder: ViteBuilder,
+  buildTasks: BuildTask[],
+): Promise<void> {
+  for (const task of buildTasks) {
+    await task.run()
+  }
+}
+
+export function resolveBuilderOptions(
+  options: BuilderOptions = {},
+): ResolvedBuilderOptions {
+  return {
+    runBuildTasks: options.runBuildTasks ?? defaultRunBuildTasks,
+  }
+}
+
+export type ResolvedBuilderOptions = Required<BuilderOptions>
+
+export interface BuilderInlineConfig extends Omit<InlineConfig, 'plugins'> {
+  plugins?: () => Plugin[]
+}
+
+export async function createViteBuilder(
+  builderOptions: BuilderOptions = {},
+  defaultBuilderInlineConfig: BuilderInlineConfig = {},
+): Promise<ViteBuilder> {
+  // Plugins passed to the Builder inline config needs to be created
+  // from a factory to ensure each build has their own instances
+  const getDefaultInlineConfig = (): InlineConfig => {
+    const { plugins } = defaultBuilderInlineConfig
+    return plugins
+      ? {
+          ...defaultBuilderInlineConfig,
+          plugins: plugins(),
+        }
+      : (defaultBuilderInlineConfig as InlineConfig)
+  }
+
+  // TODO: defineEnvironments -> config
+
+  // We resolve the whole config including plugins here but later on we
+  // need to refactor resolveConfig to only resolve the environments config
+  const defaultInlineConfig = getDefaultInlineConfig()
+  const defaultConfig = await resolveConfig(
+    defaultInlineConfig,
+    'build',
+    'production',
+    'production',
+  )
+
+  if (defaultConfig.build.lib) {
+    throw new Error('Library mode is not supported in ViteBuilder')
+  }
+  if (defaultConfig.build.watch) {
+    throw new Error('Watch mode is not yet supported in ViteBuilder')
+  }
+
+  const environments: Record<string, BuildEnvironment> = {}
+
+  const builder: ViteBuilder = {
+    environments,
+    async build() {
+      const buildTasks = []
+      for (const environment of Object.values(environments)) {
+        // We need to resolve the config again so we can properly merge options
+        // and get a new set of plugins for each build environment. The ecosystem
+        // expects plugins to be run for the same environment once they are created
+        // and to process a single bundle at a time (contrary to dev mode where
+        // plugins are built to handle multiple environments concurrently).
+
+        let userConfig = getDefaultInlineConfig()
+        const inlineConfigEnvironmentOverrides =
+          userConfig.environments?.[environment.name]
+        if (inlineConfigEnvironmentOverrides) {
+          userConfig = mergeConfig(userConfig, inlineConfigEnvironmentOverrides)
+        }
+        if (environment.config) {
+          userConfig = mergeConfig(userConfig, environment.config)
+        }
```

Comment:
`inlineConfigEnvironmentOverrides` here seems also empty.
Can this also use `defaultConfig.environment` but only merge `resolve` and `build`?

```ts
        const inlineConfigEnvironmentOverrides =
          defaultConfig.environments?.[environment.name]
        if (inlineConfigEnvironmentOverrides) {
          userConfig = mergeConfig(userConfig, {
             resolve: inlineConfigEnvironmentOverrides.resolve,
             build: inlineConfigEnvironmentOverrides.build,
          })
        }
```

Reply:
It is empty if you use the CLI, but the idea is that the user could create a Vite builder programmatically like they do with the Vite Server, and have the config inline instead of it being in a config file.

Reply:
Yeah, I understand the need of inline config for programamtical usage, but currently something is not working with `environments.(name).build` config. For example, this is what I was seeing since yesterday. Maybe this is also fixed in https://github.com/vitejs/vite/pull/16129/commits/96e825a48d8da2f3ce6238b11d16c417b1696671?

```ts
  environments: {
    server: {
      build: {
        outDir: "dist/server",  // [not ok] not included in `BuildTask.config.build`
        createEnvironment(builder, name) {
          return {
            name,
            mode: "build",
            builder,
            config: {
              build: {
                outDir: "dist/server", // [ok] I use this as a workaround currently
              },
            },
          };
        },
      },
    },
  },

```

Reply:
Would you check how it works after https://github.com/vitejs/vite/pull/16129/commits/c7e4da2e88d176a81c21781e9387b6900853cc29 ?

Reply:
Thanks for the quick update, but this issue seems to be same.

I think ideally we want to move up `environment[name].build/resolve` to the top level before doing `resolveConfig` per environment, so I was suggesting to use them as a part of `InlineConfig`. Is it possible?

https://github.com/vitejs/vite/blob/2ddf28e93b7a94835f114fcf63f22dfff92e9cca/packages/vite/src/node/build.ts#L1458-L1459

```ts
    // need to move up these two config at the top level before resolveConfig?
    defaultConfig.environments[name]?.resolve;
    defaultConfig.environments[name]?.build;

    const environmentConfig = await resolveConfig()

    // since they are not here
    environmentConfig.resolve;
    environmentConfig.build;

    // but it's still here
    environmentConfig.environments[name]?.resolve;
    environmentConfig.environments[name]?.build;
```